indexing-service/src/main/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java

```diff
@@ -0,0 +1,657 @@
+
+      // Firehose temporary directory is automatically removed when this RealtimeIndexTask completes.
+      FileUtils.forceMkdir(firehoseTempDir);
+
+      // Delay firehose connection to avoid claiming input resources while the plumber is starting up.
+      final FirehoseFactory firehoseFactory = spec.getIOConfig().getFirehoseFactory();
+      final boolean firehoseDrainableByClosing = isFirehoseDrainableByClosing(firehoseFactory);
+
+      // Skip connecting firehose if we've been stopped before we got started.
+      synchronized (this) {
+        if (!gracefullyStopped) {
+          firehose = firehoseFactory.connect(spec.getDataSchema().getParser(), firehoseTempDir);
+          committerSupplier = Committers.supplierFromFirehose(firehose);
+        }
+      }
+
+      int sequenceNumber = 0;
+      String sequenceName = makeSequenceName(getId(), sequenceNumber);
+
+      final TransactionalSegmentPublisher publisher = (segments, commitMetadata) -> {
+        final SegmentTransactionalInsertAction action = new SegmentTransactionalInsertAction(segments);
+        return toolbox.getTaskActionClient().submit(action).isSuccess();
+      };
+
+      // Time to read data!
+      while (firehose != null && (!gracefullyStopped || firehoseDrainableByClosing) && firehose.hasMore()) {
+        try {
+          InputRow inputRow = firehose.nextRow();
+
+          if (inputRow == null) {
+            log.debug("Discarded null row, considering thrownAway.");
+            metrics.incrementThrownAway();
+          } else {
+            AppenderatorDriverAddResult addResult = driver.add(inputRow, sequenceName, committerSupplier);
+
+            if (addResult.isOk()) {
+              if (addResult.isPersistRequired()) {
+                driver.persist(committerSupplier.get());
+              }
+
+              if (addResult.getNumRowsInSegment() > tuningConfig.getMaxRowsPerSegment()) {
+                publishSegments(driver, publisher, committerSupplier, sequenceName);
+
+                sequenceNumber++;
+                sequenceName = makeSequenceName(getId(), sequenceNumber);
+              }
+            } else {
+              // Failure to allocate segment puts determinism at risk, bail out to be safe.
+              // May want configurable behavior here at some point.
+              // If we allow continuing, then consider blacklisting the interval for a while to avoid constant checks.
+              throw new ISE("Could not allocate segment for row with timestamp[%s]", inputRow.getTimestamp());
+            }
+
+            metrics.incrementProcessed();
+          }
+        }
+        catch (ParseException e) {
+          if (tuningConfig.isReportParseExceptions()) {
+            throw e;
+          } else {
+            log.debug(e, "Discarded row due to exception, considering unparseable.");
+            metrics.incrementUnparseable();
+          }
+        }
+      }
+
+      if (!gracefullyStopped) {
```

Comment:
> Also, it would be great for reading code if you can split the long code into some smaller methods.

Sure, I can do that.  I had hewed to the same structure and layout as `RealtimeIndexTask` because I thought that they might be easier to review since you all are probably familiar with how that task works.  I'll refactor it.